[IA-1986] Update attached PD when updating runtime

akarukappadath opened the pull request:
1. Fix a bug in existing update disk size where we're updating boot disk size
2. Fix a bug where update disk size doesn't work because runtime is not restarted.
3. Add automation tests to validate update disk size works for gce, gce with pd, dataproc

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

akarukappadath commented:
jenkins retest

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -530,6 +532,53 @@ class RuntimeServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
+                                                           traceId: TraceId): F[Option[UpdateRuntimeMessage]] =
+    for {
+      // should machine type be updated?
+      targetMachineType <- traverseIfChanged(req.updatedMachineType, gceConfig.machineType) { mt =>
```
can you make this into a function so that it can be reused in `processUpdateGceConfigRequest`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -530,6 +532,53 @@ class RuntimeServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
+      )
+      diskOpt <- persistentDiskQuery.getById(diskId).transaction
+      disk <- F.fromEither(diskOpt.toRight(DiskNotFoundByIdException(diskId, traceId)))
+      targetDiskSize <- traverseIfChanged(req.updatedDiskSize, disk.size) { d =>
```
while you're in here, can you test to see if disk size change require restart? (there was a bug ticket that updating disk size didn't take effect until machine is restarted

akarukappadath replied:
can you elaborate on the bug? I just tried updating a disk size and it looks like it went through fine, but was the machine unable to use the disk?

Qi77Qi replied:
https://broadworkbench.atlassian.net/browse/IA-2015 is the ticket...

akarukappadath committed:
squash

test fix

bug fixes

PR feedback

test fix

manual testing done

unit test working woooo

for comp

fix things

attempt t fix unit tests

fix unit tests

remove some duplicate code

minor cleanup

fix a bug

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala:
```diff
@@ -111,6 +111,7 @@ abstract class BaseCloudServiceRuntimeMonitor[F[_]] {
             _ <- persistentDiskQuery
               .updateStatusAndIsFormatted(diskId, DiskStatus.Ready, FormattedBy.GCE, now)
               .transaction
+            _ <- persistentDiskQuery.updateNeedsResize(diskId, false).transaction
```
why is this needed?

akarukappadath replied:
once a runtime is ready (either the init or start-up script is complete) the disk will have been resized

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:
```diff
@@ -528,6 +533,65 @@ class LeoPubsubMessageSubscriber[F[_]: Timer: ContextShift](
+          .drain
+      } yield ()
+      _ <- asyncTasks.enqueue1(
+        Task(ctx.traceId, task, Some(logError(runtime.projectNameString, "restarting runtime")), ctx.now)
```
we don't have to run this async to `restartRuntime`....reason why we used asyncTaskProcessor in a main message handler is because we want this whole waiting happen async to message acknowledgement (we don't want to wait say 20 min before acking a message, cuz things get weird with pubsub in that case)...

Here the whole pollOperation is already async to the message acking, so you don't need to run ilne 556 async to `restartRuntime` anymore

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:
```diff
@@ -514,7 +515,11 @@ class LeoPubsubMessageSubscriber[F[_]: Timer: ContextShift](
+          for {
+            _ <- nowInstant.flatMap(now => persistentDiskQuery.updateSize(msg.diskId, msg.newSize, now).transaction[F])
+            _ <- persistentDiskQuery.updateNeedsResize(msg.diskId, true).transaction[F]
+            _ <- msg.restartRuntimeId.traverse(restartRuntime(_))
```
is this to account for when a `updateDisk` api called..if the disk is attached to a runtime, then we try to stop the attached runtime if it happens to be running? in other words, `restartRuntimeId` only exists when there's a runtime associated with the disk, and the runtime happens to be running?

akarukappadath replied:
`restartRuntimeId` only exists if the `UpdateDiskMessage` came from a `updateRuntime` request. If it's coming from a `updateDisk` request, `restartRuntimeId` will be `None`. The reasoning behind this is because we have the `allowStop` field in `updateRuntime,` but it doesn't make sense for the `updateDisk` endpoint. So, if someone updates the disk size through the `updateRuntime` endpoint it will restart the runtime for the user if `allowStop` is true (or throw an error if `allowStop` is false). If someone updates the disk size through the `updateDisk` endpoint, the runtime will not be restarted but the disk will have `needsResize=true` until they restart the runtime themselves

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -598,6 +599,53 @@ class RuntimeServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
+
+      // if target disk size is defined, enqueue pubsub messages to change the PD size and restart the runtime
+      _ <- targetDiskSize.traverse { size =>
+        val diskMessage = UpdateDiskMessage(diskId, size, Some(runtime.id), Some(traceId))
```
why don't we merge this intent in `UpdateRuntimeMessage` instead of a separate message for updating diskMesage? won't be get in race condition this way since `UpdateRuntimeMessage` can potentially trigger a restart as well?

akarukappadath commented:
jenkins retest

Qi77Qi commented on http/src/main/resources/init-resources/startup.sh:
```diff
@@ -154,3 +154,7 @@ if [ "$WELDER_ENABLED" == "true" ] ; then
+
+# Resize persistent disk if needed.
+echo "Resizing persistent disk attached to runtime $GOOGLE_PROJECT / $CLUSTER_NAME if disk size changed..."
+resize2fs /dev/sdb
```
this does nothing if the disk doesn't need resize...Azmina originally added `needsResize` variable, but it was added to persistent disk table, which is not the right place because we need this flag for non pd gce runtimes too. I removed `needsResize` entirely since it doesn't seem to me worth the extra complexity.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimePatchSpec.scala:
```diff
@@ -70,39 +77,132 @@ class RuntimePatchSpec extends FixtureAnyFreeSpec with LeonardoTestUtils with Le
+                  "! df -H |grep sdb"
+                )
+                .get
+              res should include("22G") //disk output is always a few more gb than what's specified
```
Really.. we specify `DiskSize(20)` and get `22G` from `df -H`?

Qi77Qi replied:
yea...not sure why

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:
```diff
@@ -369,6 +368,15 @@ object clusterQuery extends TableQuery(new ClusterTable(_)) {
     }
   }
 
+  def getActiveClusterRecordByName(project: GoogleProject,
```
Curious why we need this instead of `getActiveClusterByNameMinimal` which seems to be equivalent.. our usual practice is to not return `*Record` objects outside of DB code

Qi77Qi replied:
I added this because originally we were using `getActiveClusterByNameMinimal`, which stubs out fields outside cluster table. I didn't want to keep using `getActiveClusterByNameMinimal` because this caused a bug where I was using dataproc instances from `Runtime` object, but because it was stubbed out, it wasn't doing what I expected it's doing. By using  `getActiveClusterRecordByName`, the return type is exactly what it is, and there's no chance that caller might accidentally be using a stubbed out field

rtitle reviewed (approved):
LGTM after merge conflicts are addressed

Qi77Qi committed:
conflicts
